### PR TITLE
prevent sets with empty range

### PR DESCRIPTION
**Pull request by Menduist**
Currently,
```
echo {5..0}
```
Compiles, even though empty range should not be allowed.
This is error prone, as demonstrated by the typo in `compiler/sigmatch.nim`

This commit blocks empty range when possible (ie, when params are const-able) in sets constructions.

### CODE REVIEW

File: compiler/semexprs.nim [hunk 1]
```diff
@@ -2410,12 +2411,19 @@ proc semSetConstr(c: PContext, n: PNode): PNode =
     for i in 0..<n.len:
       if isRange(n[i]):
         checkSonsLen(n[i], 3, c.config)
-        n[i][1] = semExprWithType(c, n[i][1])
-        n[i][2] = semExprWithType(c, n[i][2])
+
+        for x in 1..2:
+          if hasUnresolvedArgs(c, n[i][x]):
```
Comment (Menduist) [inline compiler/semexprs.nim]: Not a good condition

**Comment by konsumlamm** [comment]
I find it weird to make this an error if the bounds are known, since there are legitimate use cases when the bound aren't known. So if anything, I'd make this a warning and not a hard error.

EDIT: There are probably even legitimate use cases of empty ranges when the bounds are known at compile time, but are dependent on some flags.

File: compiler/sigmatch.nim [hunk 2]
```diff
@@ -409,7 +409,7 @@ proc handleRange(f, a: PType, min, max: TTypeKind): TTypeRelation =
     elif a.kind == tyRange and
       # Make sure the conversion happens between types w/ same signedness
       (f.kind in {tyInt..tyInt64} and a[0].kind in {tyInt..tyInt64} or
-       f.kind in {tyUInt8..tyUInt32} and a[0].kind in {tyUInt8..tyInt32}) and
+       f.kind in {tyUInt8..tyUInt32} and a[0].kind in {tyUInt8..tyUInt32}) and
```
Comment (timotheecour) [inline compiler/sigmatch.nim]: is this a bugfix? if so, it should be in its own PR because the proposal in this PR is controversial, also, please provide a test case where it'd fail before and work after

File: tests/sets/tsets.nim [hunk 3]
```diff
@@ -218,3 +218,14 @@ type Foo = enum
 
 let x = { Foo1, Foo2 }
 # bug #8425
+
+# empty range should not compile
+assert not compiles({hintUser..errUser})
```
Comment (timotheecour) [inline tests/sets/tsets.nim]: tests use doAssert (see contributing.html)

**Comment by timotheecour** [comment]
I think this is a good idea that can prevent bugs but it needs more thought.
* we should support this syntax `{a..<b}` which currently doesn't compile:
```nim
when true:
  let a1 = 1..<3 # ok
  for a2 in 1..<3: discard # ok
  # type A3 = range[1..<3] # Error: range types need to be constructed with '..', '..<' is not supported
  # var a4: array[1..<3, string] # Error: ordinal type expected
  # let a5 = {1..<3} # Error: ordinal type expected
```
* in all the above cases, `a..<a` (ie, an empty range) should be legal; it's a legitimate corner case
* {a..b} where b < a may be ok to become illegal; this may require a change of implementation of `..<`
* a possible approach would be to allow a..b where b = a-1, which would avoid having to special case `..<`, yet still catch most bugs
* needs refactoring with existing `localError(c.config, n.info, "range is empty")`

* are you also handling this case?
```nim
case x
of a..b # error when b<a
else discard
```

* the `tryConstExpr` approach seems hacky
* we probably instead should issue a `rangeChecks` error for things like `a..b` where b<a are known at RT

File: compiler/sigmatch.nim (hunk 2 above)
Comment (Menduist) [reply compiler/sigmatch.nim]: Ok, I'll move it. Honestly, I saw the typo and fixed it without giving it too much thought, I'll try to find a test case

**Comment by Menduist** [comment]
> * we should support this syntax `{a..<b}` which currently doesn't compile:

Yeah, seems to be a bit of work but should be doable. Maybe the different types of range could use a little unification

> * are you also handling this case?

I think this case was already handled by the compiler, and a & b must be known at compile time, so not much issue here.

> * the `tryConstExpr` approach seems hacky

Yes, absolutely. I'm new to the compiler, couldn't find a best approach :/ I'll try to find a better way

> * we probably instead should issue a `rangeChecks` error for things like `a..b` where b<a are known at RT

Yes, it would be more consistent. I think if someone wants to do and empty set voluntarily, he should simply do `{}`, and if he has unknowns A & B for `{A..B}`, he can always check if B > A, like that :
```
proc checkXUnderY(x, y: int): bool=
  var setCheck = {}
  if y > 0:
    setCheck = {0..y}
  result = x in setCheck
```
(really dumb example but you get the point)
It's cumbersome, but I think this kind of use cases are pretty rare.

I'll switch to draft since this is a bigger PR than expected

File: compiler/sigmatch.nim (hunk 2 above)
Comment (Menduist) [reply compiler/sigmatch.nim]: Moved to #17754 with a small test case :)
I'll keep the commit in this PR while the new one is not merged

Comment (timotheecour) [reply compiler/sigmatch.nim]: @Menduist you can rebase now that it was merged :)

**Comment by arnetheduck** [comment]
https://github.com/nim-lang/Nim/issues/16492 is related because it's all part of the same logic, which at least should be internally consistent.

That said, being restrictive here removes legitimate use cases with generic code where the range is known at compile time _but not to the author of the code with the set_ - this would create (yet another) difference between compile time and runtime which makes it very difficult to use compile-time evaluated code for anything.

**Comment by Menduist** [comment]
Yeah. Having fiddled a bit with this, RT RangeDefect seems like a bad idea.

I also tried to unify the different range "types", but it's a big job that would require it's own PR.

So in the end, the goal of this PR was simply to avoid typos like #17754, and I think it does that pretty well.

> That said, being restrictive here removes legitimate use cases with generic code where the range is known at compile time but not to the author of the code with the set

As far as I can tell, the only case where this could happen is if the set of the author is in a template, so I think it's reasonable

We could narrow this check down to ordinal values, this way, any computation would bypass it. This would also remove the need to use `tryConstExpr`. Or switch back to a warning for more flexibility.
